rustic-cargo.el

```diff
@@ -458,13 +466,69 @@ If running with prefix command `C-u', read whole command from minibuffer."
                              'compile-history)))))
     (rustic-run-cargo-command command (list :mode 'rustic-cargo-run-mode))))
 
-(define-derived-mode rustic-cargo-run-mode rustic-compilation-mode "Cargo run"
-  "Mode for 'cargo run' that derives from `rustic-compilation-mode', but uses
-the keymap of `comint-mode' so user input is possible."
+(defun rustic-cargo-run-mode ()
+  (interactive)
+  (if rustic-cargo-run-use-comint
+      ;; rustic-cargo-comint-run-mode toggles the mode; we want to
+      ;; always enable.
+      (unless (and (boundp 'polymode-mode)
+                   polymode-mode
+                   (memq major-mode '(rustic-cargo-plain-run-mode
+                                      comint-mode)))
+        (rustic-cargo-comint-run-mode))
+    (rustic-cargo-plain-run-mode)))
+
+(define-derived-mode rustic-cargo-plain-run-mode rustic-compilation-mode "Cargo run"
+  "Mode for 'cargo run' that derives from `rustic-compilation-mode'.
+
+To send input to the compiled program, use
+`rustic-compile-send-input'.  If you set
+`rustic-cargo-run-use-comint' to t, you can also just type in a
+string and hit RET to send it to the program.  The latter
+approach requires installing polymode."
   (buffer-disable-undo)
   (setq buffer-read-only nil)
   (use-local-map comint-mode-map))
 
+(defun rustic-cargo-comint-run-mode ()
+  "Mode for 'cargo run' that combines `rustic-compilation-mode' with `comint-mode',
+the former for highlighting and interacting with compiler errors,
+and the latter for interacting with the compiled program."
+  ;; First time around, define the mode and invoke it.  Next time, the
+  ;; symbol will have been overwritten so this runs only once.
+  (require 'polymode)
+  (let ((docstr (documentation 'rustic-cargo-comint-run-mode)))
+    (define-hostmode poly-rustic-cargo-compilation-hostmode
+      :mode 'rustic-cargo-plain-run-mode)
+    (define-innermode poly-rustic-cargo-comint-innermode
+      :mode 'comint-mode
+      :head-matcher "^ *Running `.+`$"
+      :head-mode 'host
+      :tail-matcher "\\'"
+      :tail-mode 'host)
+    (define-polymode rustic-cargo-comint-run-mode
+      :hostmode 'poly-rustic-cargo-compilation-hostmode
+      :innermodes '(poly-rustic-cargo-comint-innermode)
+      :switch-buffer-functions '(poly-rustic-cargo-comint-switch-buffer-hook)
```

Comment:
I think it could be problematic to run these polymode commands every time `rustic-cargo-comint-run-mode` gets called.

Maybe we should check if the polymode modes are already defined.

Reply:
That's already taken care of.  This code runs only once, when `rustic-cargo-comint-run-mode` is called for the first time.  That symbol gets overwritten with `define-polymode`, so from the second time on, the already-defined mode is called.